etcdserver/server.go

```diff
@@ -408,7 +408,9 @@ func NewServer(cfg *ServerConfig) (srv *EtcdServer, err error) {
 	srv.kv = mvcc.New(srv.be, srv.lessor, &srv.consistIndex)
 	if beExist {
 		kvindex := srv.kv.ConsistentIndex()
-		if snapshot != nil && kvindex < snapshot.Metadata.Index {
+		// TODO: remove kvindex != 0 checking when we do not expect users to upgrade
+		// etcd from pre-3.0 release.
+		if snapshot != nil && kvindex != 0 && kvindex < snapshot.Metadata.Index {
```

Comment:
add a warning just in case?
```
if snapshot ... { 
    if kvindex != 0 { return ... }
    warning("consistent index never saved but snapshot.Metadata.Index=%d", snapshot.Metadata.Index)
}
```